definitions/npm/react-redux_v7.x.x/flow_v0.142.x-/react-redux_v7.x.x.js

```diff
@@ -200,7 +200,15 @@ declare module "react-redux" {
   // Typings for Hooks
   // ------------------------------------------------------------
 
-  declare export function useDispatch<D>(): D;
+  declare export function useDispatch<D>(): (
+    & (<T: { [key: string]: any }>(T) => T)
+    // Supports thunks at their various lengths and use cases depending if user has typed them as tuple vs array
+    & (<T>((...args: [any]) => T) => T)
+    & (<T>((...args: [any, any]) => T) => T)
+    & (<T>((...args: [any, any, any]) => T) => T)
+    & (<T>((...args: Array<any>) => T) => T)
+    & D
```

Comment:
I'm surprised an intersection works here, isn't it an intersection instead? Or I could use a bit of explanations TBH.

Reply:
So `useDispatch` is a function that returns a function, the returned function needs to be an intersection so it can accept a wide variety of api contracts depending on its usage and have them all be valid.
It's not a union type because if it was, there would need to be a type refinement before usage of the function